src/cross_section/src/cross_section.cpp

```diff
@@ -0,0 +1,269 @@
+// Copyright 2023 The Manifold Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "cross_section.h"
+
+#include <clipper2/clipper.h>
+
+#include <vector>
+
+#include "clipper2/clipper.core.h"
+#include "clipper2/clipper.engine.h"
+#include "clipper2/clipper.offset.h"
+#include "glm/ext/vector_float2.hpp"
+#include "public.h"
+
+using namespace manifold;
+
+namespace {
+const int precision_ = 8;
+
+C2::ClipType cliptype_of_op(CrossSection::OpType op) {
+  C2::ClipType ct = C2::ClipType::Union;
+  switch (op) {
+    case CrossSection::OpType::Add:
+      break;
+    case CrossSection::OpType::Subtract:
+      ct = C2::ClipType::Difference;
+      break;
+    case CrossSection::OpType::Intersect:
+      ct = C2::ClipType::Intersection;
+      break;
+    case CrossSection::OpType::Xor:
+      ct = C2::ClipType::Xor;
+      break;
+  };
+  return ct;
+}
+
+C2::JoinType jt(CrossSection::JoinType jointype) {
+  C2::JoinType jt = C2::JoinType::Square;
+  switch (jointype) {
+    case CrossSection::JoinType::Square:
+      break;
+    case CrossSection::JoinType::Round:
+      jt = C2::JoinType::Round;
+      break;
+    case CrossSection::JoinType::Miter:
+      jt = C2::JoinType::Miter;
+      break;
+  };
+  return jt;
+}
+
+C2::PathD pathd_of_contour(const SimplePolygon& ctr) {
+  auto p = C2::PathD(ctr.size());
+  for (auto v : ctr) {
+    p.push_back(C2::PointD(v.x, v.y));
+  }
+  return p;
+}
+}  // namespace
+
+namespace manifold {
+CrossSection::CrossSection() { paths_ = C2::PathsD(); }
+CrossSection::~CrossSection() = default;
+CrossSection::CrossSection(CrossSection&&) noexcept = default;
+CrossSection& CrossSection::operator=(CrossSection&&) noexcept = default;
+CrossSection::CrossSection(const CrossSection& other) {
+  paths_ = C2::PathsD(other.paths_);
+}
+CrossSection::CrossSection(C2::PathsD ps) { paths_ = ps; }
+
+CrossSection::CrossSection(const SimplePolygon& contour) {
+  auto ps = C2::PathsD{(pathd_of_contour(contour))};
+  paths_ = C2::Union(ps, C2::FillRule::Positive);
+}
+
+CrossSection::CrossSection(const Polygons& contours) {
+  auto ps = C2::PathsD(contours.size());
+  for (auto ctr : contours) {
+    ps.push_back(pathd_of_contour(ctr));
+  }
+  paths_ = C2::Union(ps, C2::FillRule::Positive);
```

Comment:
I suppose the quick test is to check if removing this line fixes the tests. Still Revolve and Extrude are getting the wrong genus (!) and some of the tests are off by a lot in volume. I think there's a real issue somewhere.

Reply:
Coming back with slightly fresher eyes, I caught that I missed a couple fixes earlier with how I was incorrectly using the `std::vector` constructor to pre-allocate. With those switched to `reserve`, the empty padding on the front is gone, and now only two failures remain.

Boolean.Coplanar and Boolean-CoplanarProp both have lose a small number of vertices/triangles.

Reply:
Good finds! And it gives me further confidence in our testing regimen. Speaking of, the codecov report does give some pointers now and then about where some extra tests would be helpful. As you can probably tell, I tend to prefer end-to-end tests over unit tests - test more with less code whenever possible!